service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -197,27 +200,39 @@ public final TenantClientFactory getTenantClientFactory() {
      */
     @Qualifier(DeviceConnectionConstants.DEVICE_CONNECTION_ENDPOINT)
     @Autowired
-    public final void setDeviceConnectionClientFactory(final DeviceConnectionClientFactory factory) {
+    public final void setDeviceConnectionClientFactory(final BasicDeviceConnectionClientFactory factory) {
         this.deviceConnectionClientFactory = Objects.requireNonNull(factory);
     }
 
     /**
      * Gets the factory used for creating a client for the Device Connection service.
      *
      * @return The factory.
+     * @deprecated Use {@link #getDeviceConnectionClient(String)} in order to access
+     *             device connection information.
      */
+    @Deprecated(forRemoval = true)
     public final DeviceConnectionClientFactory getDeviceConnectionClientFactory() {
-        return deviceConnectionClientFactory;
+        if (deviceConnectionClientFactory instanceof DeviceConnectionClientFactory) {
+            return (DeviceConnectionClientFactory) deviceConnectionClientFactory;
+        } else {
+             return null;
+        }
     }
 
     /**
      * Gets a client for interacting with the Device Connection service.
      *
      * @param tenantId The tenant that the client is scoped to.
      * @return The client.
+     * @throws IllegalStateException if no client factory is set.
      */
     protected final Future<DeviceConnectionClient> getDeviceConnectionClient(final String tenantId) {
-        return getDeviceConnectionClientFactory().getOrCreateDeviceConnectionClient(tenantId);
+
+        if (deviceConnectionClientFactory == null) {
```

Comment:
When could this be `null`? Before `startInternal()` is called?

Reply:
yes. I do not know of any other formal way of expressing that setting this property is required for this object functioning properly ...